### PR TITLE
*: make assertion check more reasonable

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Make sure there are no "ASSERTION fail" logs in `session` and `executor` package unit test

### What is changed and how it works?

1. Add a `ConfirmAssertions()` function to handle the StmtRollback case:

create table t (a, b unique key)
insert into t values (1, 1)
insert into t values (2, 1)  // unique key conflict, statement fail, should not add assertion

2. Add check assertion None to update operation

insert into t values (3)
update t set i = 4 where i = 3   // index 3 exist  
update t set i = 5 where i = 4   // index 4 exist 

In this case, "index 4 exist" is not a reasonable assertion,
we can add a "index 4 none" assertion in the statement "update t set i = 4 where i = 3" to invalidate "index 4 exist"

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
Make sure no "ASSERTION fail" in the log

### CODE REVIEW

File: table/index.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ type Index interface {
 	// Meta returns IndexInfo.
 	Meta() *model.IndexInfo
 	// Create supports insert into statement.
-	Create(ctx sessionctx.Context, rm kv.RetrieverMutator, indexedValues []types.Datum, h int64, opts ...*CreateIdxOpt) (int64, error)
```
Comment (jackysp) [inline table/index.go]: There are some options for `Create`, it is better to make `ss kv.Transaction` as an option.

**Comment by tiancaiamao** [comment]
I'll update this PR after https://github.com/pingcap/tidb/pull/11018 is merged

File: table/index.go (hunk 1 above)
Comment (tiancaiamao) [reply table/index.go]: Updated

**Comment by tiancaiamao** [comment]
PTAL @jackysp @lysu @coocood

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 2]
```diff
@@ -643,6 +643,7 @@ func checkConflictValue(iter *Iterator, m *kvrpcpb.Mutation, startTS uint64) err
 	if !ok {
```
Comment (coocood) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: Need to check Assertion_Exist here?

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 3]
```diff
@@ -722,12 +729,6 @@ func prewriteMutation(db *leveldb.DB, batch *leveldb.Batch, mutation *kvrpcpb.Mu
 		return errors.Trace(err)
 	}
 
-	// Check assertions.
-	if (ok && mutation.Assertion == kvrpcpb.Assertion_NotExist) ||
-		(!ok && mutation.Assertion == kvrpcpb.Assertion_Exist) {
```
Comment (lysu) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: should this method still need check `ok && mutation.Assertion == kvrpcpb.Assertion_NotExist` ?

it seems checks inside `checkConflictValue` only cover `!ok` ?

Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: ok && mutation.Assertion == kvrpcpb.Assertion_NotExist

!ok && mutation.Assertion == kvrpcpb.Assertion_Exist

Now split into two different cases, the check is basically the same.

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 2 above)
Comment (tiancaiamao) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: Done

**Comment by lysu** [comment]
LGTM

**Comment by jackysp** [comment]
PTAL @coocood

**Pull request merged**